File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> lastSeen = new HashMap();

        int left = 0;
        int maxLength = 0;

        for(int right = 0; right < s.length(); right++) {
            if (lastSeen.containsKey(s.charAt(right)) && left <= lastSeen.get(s.charAt(right))) {
                left = lastSeen.get(s.charAt(right)) + 1;
            } 

            maxLength = Math.max(maxLength, right - left + 1);
            lastSeen.put(s.charAt(right), right);
        }

        return maxLength;
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        lastSeen = {}

        left, maxLen = 0, 0

        for right in range(len(s)):
            char = s[right]
            if char in lastSeen and lastSeen[char] >= left:
                left = lastSeen[char] + 1
            
            lastSeen[char] = right

            maxLen = max(maxLen, right - left + 1)

        return maxLen

        